src/circuit/gadget/sinsemilla/chip.rs

```diff
@@ -0,0 +1,318 @@
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub struct SinsemillaConfig<const K: usize, const MAX_WORDS: usize> {
+    q_sinsemilla1: Selector,
+    q_sinsemilla2: Selector,
+    word: Column<Advice>,
+    x_a: Column<Advice>,
+    x_p: Column<Advice>,
+    lambda: (Column<Advice>, Column<Advice>),
+    perm: Permutation,
+    generator_table: GeneratorTableConfig<pallas::Affine, K>,
+}
+
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub struct SinsemillaChip<const K: usize, const MAX_WORDS: usize> {
+    config: SinsemillaConfig<K, MAX_WORDS>,
+}
+
+impl<const K: usize, const MAX_WORDS: usize> Chip<pallas::Base> for SinsemillaChip<K, MAX_WORDS> {
+    type Config = SinsemillaConfig<K, MAX_WORDS>;
+    type Loaded = ();
+
+    fn config(&self) -> &Self::Config {
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &()
+    }
+}
+
+impl<const K: usize, const MAX_WORDS: usize> SinsemillaChip<K, MAX_WORDS> {
+    pub fn construct(config: <Self as Chip<pallas::Base>>::Config) -> Self {
+        Self { config }
+    }
+
+    pub fn load(
+        config: SinsemillaConfig<K, MAX_WORDS>,
+        layouter: &mut impl Layouter<pallas::Base>,
+    ) -> Result<<Self as Chip<pallas::Base>>::Loaded, Error> {
+        // Load the lookup table.
+        config.generator_table.load(layouter)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn configure(
+        meta: &mut ConstraintSystem<pallas::Base>,
+        advices: [Column<Advice>; 5],
+        lookup: (Column<Fixed>, Column<Fixed>, Column<Fixed>),
+        perm: Permutation,
+    ) -> <Self as Chip<pallas::Base>>::Config {
+        // Sinsemilla selector for lookup and expr1
+        let q_sinsemilla1 = meta.selector();
+        // Sinsemilla selector for expr2
+        let q_sinsemilla2 = meta.selector();
+
+        // Rename advice columns for the Sinsemilla context
+        let word = advices[0];
+        let x_a = advices[1];
+        let x_p = advices[2];
+        let lambda = (advices[3], advices[4]);
+
+        // Generator table config
+        let generator_table =
+            GeneratorTableConfig::configure(meta, q_sinsemilla1, lookup, word, x_a, x_p, lambda);
+
+        // Create custom gates
+        meta.create_gate("Sinsemilla gate 1", |meta| {
+            // (λ_{2,i})^2 − (x_{A,i+1} + (λ_{1,i}^2 − x_{A,i} − x_{P,i}) + x_{A,i})
+            let expr1 = {
+                let q_sinsemilla1 = meta.query_selector(q_sinsemilla1);
+                let lambda1 = meta.query_advice(lambda.0, Rotation::cur());
+                let lambda1_sq = lambda1.clone() * lambda1;
+                let lambda2 = meta.query_advice(lambda.1, Rotation::cur());
+                let lambda2_sq = lambda2.clone() * lambda2;
+                let x_a_next = meta.query_advice(x_a, Rotation::next());
+                let x_p_cur = meta.query_advice(x_p, Rotation::cur());
+
+                q_sinsemilla1 * (lambda2_sq - (x_a_next + (lambda1_sq - x_p_cur)))
```

Comment:
```suggestion
                q_sinsemilla1 * (lambda2.square() - (x_a_next + (lambda1.square() - x_p_cur)))
```

Reply:
It's the degree 3 constraint here (the `x_{A,i}`'s cancel out):
![image](https://user-images.githubusercontent.com/14323671/121926056-21adc680-cd70-11eb-9a63-cb06e9666479.png)